### PR TITLE
feat: Added docTR integration for object detection

### PR DESCRIPTION
**Pull request by frgfm**
Hi there :wave: 

To follow-up on https://github.com/mindee/doctr/pull/836, I followed the instructions from the [documentation](https://huggingface.co/docs/hub/adding-a-library#set-up-the-inference-api) to make this integration!

The goal is to use an object detection model from the hub with docTR in a widget :hugs: 

There are a few things to handle on our end first:
- [x] https://github.com/mindee/doctr/pull/836
- [x] either edit the Dockerfile to support git-based installation or make a docTR release

And there is one thing I still don't know how to handle: where is the front-end widget defined? :pray:  This PR is about the back-end, but I couldn't find any reference about the expected returned format of the model so that it interfaces nicely, similarly to https://huggingface.co/facebook/detr-resnet-50

If there is anything that already looks wrong in the PR, please let me know :)

cc @osanseviero

### CODE REVIEW

**Comment by osanseviero** [comment]
Quick answers, but I can dive in more detailed if you want :smile: 

>  either edit the Dockerfile to support git-based installation or make a docTR release

We suggest doing a release, but if that is prevented by any reason, you can also pip install from a git commit such as https://github.com/huggingface/huggingface_hub/blob/main/api-inference-community/docker_images/fairseq/requirements.txt#L8, so this should not be a blocker. 

> And there is one thing I still don't know how to handle: where is the front-end widget defined? pray This PR is about the back-end, but I couldn't find any reference about the expected returned format of the model so that it interfaces nicely, similarly to https://huggingface.co/facebook/detr-resnet-50

As long as the backend meets the expected input/output, the widgets/frontend will work out of the box for you. Right now we have not implemented object detection in community API, but we can for sure add it - it should not be significant work from our side - we can prioritize in #332 since this is the first time this has been requested.

File: api-inference-community/docker_images/doctr/tests/test_api_object_detection.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        self.assertEqual(
+            response.status_code,
+            200,
+        )
+        content = json.loads(response.content)
+        self.assertEqual(type(content), list)
+        self.assertEqual(set(type(el) for el in content), {dict})
+        self.assertEqual(
+            set((k, type(v)) for el in content for (k, v) in el.items()),
+            {("label", str), ("score", float), ("box", Tuple[float, float, float, float])},
```
Comment (osanseviero) [inline api-inference-community/docker_images/doctr/tests/test_api_object_detection.py:69]: To fix with the format change for the box

File: api-inference-community/tests/test_dockers.py [hunk 2]
```diff
@@ -240,6 +240,10 @@ def test_keras(self):
             "keras", "image-classification", "nateraw/keras-mnist-convnet"
         )
 
+    def test_doctr(self):
```
Comment (osanseviero) [inline api-inference-community/tests/test_dockers.py:243]: Please also add a GH Actions workflow to test this image. You can just duplicate this and do some minor modifications https://github.com/huggingface/huggingface_hub/blob/main/.github/workflows/python-api-stanza.yaml

**Review by osanseviero — COMMENTED** [review]
Very nice PR! Thanks for this :rocket: 

Please let us know if you have any feedback on anything that could be simpler from the development perspective, since it's always appreciated.

I left a couple of comments and also opened #749 which should allow you to test. Feel free to install `huggingface_hub` from there to test locally. Once addressed, I think we can move ahead :rocket:

**Review by Narsil — COMMENTED** [review]
This is a great looking PR.

I agree with @osanseviero 's comments.

We can also take care of the minor modifications if you want after we enable `object-detection`.

Thank you very much for this !

File: api-inference-community/docker_images/doctr/app/pipelines/object_detection.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+from PIL import Image
+import torch
+from torchvision.transforms import Compose, ConvertImageDtype, PILToTensor
+from doctr.models.obj_detection.factory import from_hub
+
+
+class ObjectDetectionPipeline(Pipeline):
+    def __init__(self, model_id: str):
+
+        self.model = from_hub(model_id).eval()
```
[the remark this answers is not in the document]
Comment (frgfm) [reply api-inference-community/docker_images/doctr/app/pipelines/object_detection.py:14]: Yes! It's instantiating the model and loading the parameters (map_location="cpu")

File: api-inference-community/docker_images/doctr/app/pipelines/object_detection.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        im = inputs.convert("RGB")
+        inputs = self.transform(im).unsqueeze(0)
+
+        with torch.inference_mode():
+            out = self.model(inputs)[0]
+
+        return [
+            {"label": self.labels[idx], "score": score.item(), "box": tuple(box.numpy().tolist())}
+            for idx, score, box in zip(out['labels'], out['scores'], out['boxes'])
+        ]
```
[the remark this answers is not in the document]
Comment (frgfm) [reply api-inference-community/docker_images/doctr/app/pipelines/object_detection.py:44]: Ah thank you! That's what I was looking for :pray:

File: api-inference-community/docker_images/doctr/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,5 @@
+starlette==0.14.2
+api-inference-community==0.0.19
+pillow==8.4.0
+-e git+https://github.com/mindee/doctr.git#egg=python-doctr[torch]
+huggingface-hub
```
[the remark this answers is not in the document]
Comment (frgfm) [reply api-inference-community/docker_images/doctr/requirements.txt:5]: Correct, I'll remove this!

**Comment by frgfm** [comment]
Thanks for the reviews @osanseviero @Narsil :pray: 

I just updated the PR accordingly! I'd like to also use what you've done in #749 and related PRs, but I must admit I'm not sure how :sweat_smile: How can I test this as you suggested?

File: api-inference-community/docker_images/doctr/app/pipelines/object_detection.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (osanseviero) [reply api-inference-community/docker_images/doctr/app/pipelines/object_detection.py:14]: Awesome 🚀

**Comment by osanseviero** [comment]
No worries, it can be a bit complicated. @Narsil would it be possible to have a new release so we can specify that version here?

In the meantime, @frgfm, you could do https://github.com/huggingface/huggingface_hub/tree/main/api-inference-community#modifying-files-within-api-inference-communityroutesvalidationpy to test things out. 

Another option to run your tests is to do pip install from head. That is, update your repo by merging from main, moving to `api-inference-community` directory, and then installing with `pip install -e .`

**Comment by osanseviero** [comment]
@frgfm with #755 we have a new release and you can use `api-inference-community` at version `0.0.21`. Installing from there and updating the requirements should fix the issues and allow you to test 🚀

**Comment by Narsil** [comment]
Yes, there's no authentication on this part of the code.

All authentication is taken care of in our internal routings beforehand, so you don't have to worry about it.

**Comment by Narsil** [comment]
No need to remove `Pillow` just wait a minute I am investigating the root cause, your file could be ok and the root cause somewhere else.

**Comment by frgfm** [comment]
> Yes, there's no authentication on this part of the code.
> 
> All authentication is taken care of in our internal routings beforehand, so you don't have to worry about it.

Ok, I'm still having trouble testing it :sweat: 
The webserver is running on port 8000, so I tried:
```python
import json

import requests

API_URL = "https://localhost:8000"

def query(filename):
    with open(filename, "rb") as f:
        data = f.read()
    response = requests.request("POST", API_URL, headers=headers, data=data)
    return json.loads(response.content.decode("utf-8"))

data = query("cats.jpg")
 ```

and I also tried:
```python
"https://localhost:8000/models/mindee/fasterrcnn_mobilenet_v3_large_fpn"
```

and I always get 
```
---------------------------------------------------------------------------
SSLEOFError                               Traceback (most recent call last)
~/miniconda3/lib/python3.8/site-packages/urllib3/connectionpool.py in urlopen(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)
    698             # Make the request on the httplib connection object.
--> 699             httplib_response = self._make_request(
    700                 conn,

~/miniconda3/lib/python3.8/site-packages/urllib3/connectionpool.py in _make_request(self, conn, method, url, timeout, chunked, **httplib_request_kw)
    381         try:
--> 382             self._validate_conn(conn)
    383         except (SocketTimeout, BaseSSLError) as e:

~/miniconda3/lib/python3.8/site-packages/urllib3/connectionpool.py in _validate_conn(self, conn)
   1009         if not getattr(conn, "sock", None):  # AppEngine might not have  `.sock`
-> 1010             conn.connect()
   1011 

~/miniconda3/lib/python3.8/site-packages/urllib3/connection.py in connect(self)
    410 
--> 411         self.sock = ssl_wrap_socket(
    412             sock=conn,

~/miniconda3/lib/python3.8/site-packages/urllib3/util/ssl_.py in ssl_wrap_socket(sock, keyfile, certfile, cert_reqs, ca_certs, server_hostname, ssl_version, ciphers, ssl_context, ca_cert_dir, key_password, ca_cert_data, tls_in_tls)
    448     if send_sni:
--> 449         ssl_sock = _ssl_wrap_socket_impl(
    450             sock, context, tls_in_tls, server_hostname=server_hostname

~/miniconda3/lib/python3.8/site-packages/urllib3/util/ssl_.py in _ssl_wrap_socket_impl(sock, ssl_context, tls_in_tls, server_hostname)
    492     if server_hostname:
--> 493         return ssl_context.wrap_socket(sock, server_hostname=server_hostname)
    494     else:

~/miniconda3/lib/python3.8/ssl.py in wrap_socket(self, sock, server_side, do_handshake_on_connect, suppress_ragged_eofs, server_hostname, session)
    499         # ctx._wrap_socket()
--> 500         return self.sslsocket_class._create(
    501             sock=sock,

~/miniconda3/lib/python3.8/ssl.py in _create(cls, sock, server_side, do_handshake_on_connect, suppress_ragged_eofs, server_hostname, context, session)
   1039                         raise ValueError("do_handshake_on_connect should not be specified for non-blocking sockets")
-> 1040                     self.do_handshake()
   1041             except (OSError, ValueError):

~/miniconda3/lib/python3.8/ssl.py in do_handshake(self, block)
   1308                 self.settimeout(None)
-> 1309             self._sslobj.do_handshake()
   1310         finally:

SSLEOFError: EOF occurred in violation of protocol (_ssl.c:1131)

During handling of the above exception, another exception occurred:

MaxRetryError                             Traceback (most recent call last)
~/miniconda3/lib/python3.8/site-packages/requests/adapters.py in send(self, request, stream, timeout, verify, cert, proxies)
    438             if not chunked:
--> 439                 resp = conn.urlopen(
    440                     method=request.method,

~/miniconda3/lib/python3.8/site-packages/urllib3/connectionpool.py in urlopen(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)
    754 
--> 755             retries = retries.increment(
    756                 method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]

~/miniconda3/lib/python3.8/site-packages/urllib3/util/retry.py in increment(self, method, url, response, error, _pool, _stacktrace)
    573         if new_retry.is_exhausted():
--> 574             raise MaxRetryError(_pool, url, error or ResponseError(cause))
    575 

MaxRetryError: HTTPSConnectionPool(host='localhost', port=8000): Max retries exceeded with url: / (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1131)')))

During handling of the above exception, another exception occurred:

SSLError                                  Traceback (most recent call last)
<ipython-input-2-a4b68c07dcb7> in <module>
     11     return json.loads(response.content.decode("utf-8"))
     12 
---> 13 data = query("/home/fg/Downloads/testset_artefact_detection/testset/1138-receipt.jpg")

<ipython-input-2-a4b68c07dcb7> in query(filename)
      8     with open(filename, "rb") as f:
      9         data = f.read()
---> 10     response = requests.request("POST", API_URL, data=data)
     11     return json.loads(response.content.decode("utf-8"))
     12 

~/miniconda3/lib/python3.8/site-packages/requests/api.py in request(method, url, **kwargs)
     59     # cases, and look like a memory leak in others.
     60     with sessions.Session() as session:
---> 61         return session.request(method=method, url=url, **kwargs)
     62 
     63 

~/miniconda3/lib/python3.8/site-packages/requests/sessions.py in request(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)
    540         }
    541         send_kwargs.update(settings)
--> 542         resp = self.send(prep, **send_kwargs)
    543 
    544         return resp

~/miniconda3/lib/python3.8/site-packages/requests/sessions.py in send(self, request, **kwargs)
    653 
    654         # Send the request
--> 655         r = adapter.send(request, **kwargs)
    656 
    657         # Total elapsed time of the request (approximately)

~/miniconda3/lib/python3.8/site-packages/requests/adapters.py in send(self, request, stream, timeout, verify, cert, proxies)
    512             if isinstance(e.reason, _SSLError):
    513                 # This branch is for urllib3 v1.22 and later.
--> 514                 raise SSLError(e, request=request)
    515 
    516             raise ConnectionError(e, request=request)

SSLError: HTTPSConnectionPool(host='localhost', port=8000): Max retries exceeded with url: / (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1131)')))
```

Could you be a bit more specific on how I'm supposed to test the route please? :pray: 

About the docker, when I build it locally, I'm still having a weird issue with the API inference lib having a strict version of pillow:
```
The conflict is caused by:
    api-inference-community 0.0.21 depends on pillow==8.2.0
    python-doctr[torch] 0.5.1a0 depends on Pillow>=8.3.2
```

I'm updating my forked by merging upstream main to investigate

File: api-inference-community/docker_images/doctr/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,30 @@
+FROM tiangolo/uvicorn-gunicorn:python3.8
+LABEL maintainer="me <me@example.com>"
+
+# Add any system dependency here
+RUN apt-get update -y && apt-get install libgl1-mesa-glx -y
```
Comment (Narsil) [inline api-inference-community/docker_images/doctr/Dockerfile:5]: This is needed by `cv2`

File: api-inference-community/docker_images/doctr/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+
+# Most DL models are quite large in terms of memory, using workers is a HUGE
+# slowdown because of the fork and GIL with python.
+# Using multiple pods seems like a better default strategy.
+# Feel free to override if it does not make sense for your library.
+ARG max_workers=1
+ENV MAX_WORKERS=$max_workers
+ENV HUGGINGFACE_HUB_CACHE=/data
+ENV TORCH_HOME=/data/torch_hub/
```
Comment (Narsil) [inline api-inference-community/docker_images/doctr/Dockerfile:19]: This is needed to get caching in place.

**Comment by Narsil** [comment]
As for the workflow. 

I used roughly what is described here: 
https://github.com/huggingface/huggingface_hub/tree/main/api-inference-community

```
./manage.py docker --model-id=mindee/fasterrcnn_mobilenet_v3_large_fpn --framework=doctr
```
In order to get the framework we will need to update the README.md here https://huggingface.co/mindee/fasterrcnn_mobilenet_v3_large_fpn/blob/main/README.md
to include `framework: doctr` in the header (between `---`)


Launching that command enabled me to figure out that we indeed missed some updates to be more accomodable (datasets==1.10.0) was the biggest culprit, and it's sometimes hard to see because pip will iteratively downgrade all packages in an attempt to find a resolution, which in this case isn't a Pillow issue, but a datasets issue, which in turn forced a downgrade of `api-inference-community` and then lead to a final (for pip) conflict of pillow.

This is tedious to follow/fix, so that's why I took the liberty to fix it.

then once the docker works

```
curl -X POST --data-binary "@plane.jpg" http://localhost:8000
```
`plane.jpg` must be a file in the local directory to send to the API.

**Pull request merged**